core/lib/vm_interface/src/types/tracer.rs

```diff
@@ -104,6 +120,30 @@ pub enum ValidationError {
     ViolatedRule(ViolatedValidationRule),
 }
 
+/// Traces the validation of a transaction, providing visibility into the aspects the transaction interacts with.
+/// For instance, the `timestamp_asserter_range` represent the range within which the transaction might make
+/// assertions on `block.timestamp`. This information is crucial for the caller, as expired transactions should
+/// be excluded from the mempool.
+#[derive(Debug, Clone, Default)]
+pub struct ValidationTraces {
+    pub timestamp_asserter_range: Option<Range<u64>>,
+}
+
+impl ValidationTraces {
+    /// Merges two ranges by selecting the maximum of the start values and the minimum of the end values,
+    /// producing the narrowest possible time window. Note that overlapping ranges are essential;
+    /// a lack of overlap would have triggered an assertion failure in the `TimestampAsserter` contract,
+    /// as `block.timestamp` cannot satisfy two non-overlapping ranges.
+    pub fn apply_range(&mut self, new_range: Range<u64>) {
```

Comment:
Bikeshedding: If we plan to add new fields to `ValidationTraces` in the future, this method name / docs look too ambiguous. It would make sense to call it like `(apply|narrow)_timestamp_asserter_range`.